src/librustc_mir/build/expr/as_place.rs

```diff
@@ -309,4 +306,164 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             }
         }
     }
+
+    /// Lower an index expression
+    ///
+    /// This has two complications;
+    ///
+    /// * We need to do a bounds check.
+    /// * We need to ensure that the bounds check can't be invalidated using an
+    ///   expression like `x[1][{x = y; 2}]`. We use fake borrows here to ensure
+    ///   that this is the case.
+    fn lower_index_expression(
+        &mut self,
+        mut block: BasicBlock,
+        base: ExprRef<'tcx>,
+        index: ExprRef<'tcx>,
+        mutability: Mutability,
+        fake_borrow_temps: Option<&mut Vec<Local>>,
+        temp_lifetime: Option<region::Scope>,
+        expr_span: Span,
+        source_info: SourceInfo
+    ) -> BlockAnd<PlaceBuilder<'tcx>> {
+        let lhs = self.hir.mirror(base);
+        let (usize_ty, bool_ty) = (self.hir.usize_ty(), self.hir.bool_ty());
+
+        let base_fake_borrow_temps = &mut Vec::new();
+        let is_outermost_index = fake_borrow_temps.is_none();
+        let fake_borrow_temps = fake_borrow_temps.unwrap_or(base_fake_borrow_temps);
+
+        let place_builder = unpack!(block = self.expr_as_place(
+                    block,
+                    lhs,
+                    mutability,
+                    Some(fake_borrow_temps),
+                ));
+
+        // Making this a *fresh* temporary also means we do not have to worry
+        // about the index changing later: Nothing will ever change this
+        // temporary. The "retagging" transformation (for Stacked Borrows)
+        // relies on this.
+        let idx = unpack!(block = self.as_temp(
+            block,
+            temp_lifetime,
+            index,
+            Mutability::Not,
+        ));
+
+        let slice = place_builder.clone().into_place();
+        // bounds check:
+        let (len, lt) = (
+            self.temp(usize_ty.clone(), expr_span),
+            self.temp(bool_ty, expr_span),
+        );
+        self.cfg.push_assign(
+            block,
+            source_info, // len = len(slice)
+            &len,
+            Rvalue::Len(slice),
+        );
+        self.cfg.push_assign(
+            block,
+            source_info, // lt = idx < len
+            &lt,
+            Rvalue::BinaryOp(
+                BinOp::Lt,
+                Operand::Copy(Place::from(idx)),
+                Operand::Copy(len.clone()),
+            ),
+        );
+
+        let msg = BoundsCheck {
+            len: Operand::Move(len),
+            index: Operand::Copy(Place::from(idx)),
+        };
+        let success = self.assert(block, Operand::Move(lt), true, msg, expr_span);
```

Comment:
Could we split the bounds check stuff from the fake-read stuff into two methods to make it clearer where each bit starts & ends?